Kernel/Devices/KCOVDevice.cpp

```
@@ -0,0 +1,148 @@
+    thread_instance->remove(tid);
+}
+
+void KCOVDevice::free_process()
+{
+    auto process = Process::current();
+    auto pid = process->pid();
+
+    auto maybe_kcov_instance = proc_instance->get(pid);
+    if (!maybe_kcov_instance.has_value())
+        return;
+
+    auto kcov_instance = maybe_kcov_instance.value();
+    VERIFY(kcov_instance->state == KCOVInstance::OPENED);
+    kcov_instance->state = KCOVInstance::UNUSED;
+    proc_instance->remove(pid);
+    delete kcov_instance;
+}
+
+KResultOr<NonnullRefPtr<FileDescription>> KCOVDevice::open(int options)
+{
+    auto process = Process::current();
+    auto pid = process->pid();
+    if (proc_instance->get(pid).has_value())
+        return EBUSY; // This process already open()ed the kcov device
+    auto kcov_instance = new KCOVInstance(pid);
+    kcov_instance->state = KCOVInstance::OPENED;
+    proc_instance->set(pid, kcov_instance);
+
+    return File::open(options);
+}
+
+int KCOVDevice::ioctl(FileDescription&, unsigned request, FlatPtr arg)
+{
+    int error = 0;
+    auto thread = Thread::current();
+    auto tid = thread->tid();
+    auto pid = thread->pid();
+    auto maybe_kcov_instance = proc_instance->get(pid);
+    if (!maybe_kcov_instance.has_value())
+        return ENXIO; // This proc didn't open the kcov dev yet
+    auto kcov_instance = maybe_kcov_instance.value();
+
+    ScopedSpinLock lock(kcov_instance->lock);
+    switch (request) {
+    case KCOV_SETBUFSIZE: {
+        if (kcov_instance->state >= KCOVInstance::TRACING) {
+            error = EBUSY;
+            break;
+        }
+        error = kcov_instance->buffer_allocate(arg);
+        break;
+    }
+    case KCOV_ENABLE: {
+        if (kcov_instance->state >= KCOVInstance::TRACING) {
+            error = EBUSY;
+            break;
+        }
+        if (!kcov_instance->has_buffer()) {
+            error = ENOBUFS;
+            break;
+        }
+        VERIFY(kcov_instance->state == KCOVInstance::OPENED);
```

This allows the user (or at least root) to panic the kernel. Maybe return an error instead?

---

There is (or should at least be) no way for users to make this VERIFY evaluate true. I also can't reasonably return an error here, as this function is called from Process::die(). If we don't clean up the KCOVInstance here, we will leak it. We could remove these verifies and just go ahead and set expected state / delete indiscriminately, however I think it preferable to find out if our code is flawed in a way that would lead us to crashing on this VERIFY. Additionally this code is hidden behind the KCOV feature flag that can't even accidentally slip into a production build due to it's performance penalty.